gsl/gsl_util

```diff
@@ -135,30 +135,30 @@ inline T narrow(U u)
 // at() - Bounds-checked way of accessing static arrays, std::array, std::vector
 //
 template <class T, size_t N>
-constexpr T& at(T (&arr)[N], size_t index)
+constexpr T& at(T (&arr)[N], std::ptrdiff_t index)
 {
-    Expects(index < N);
-    return arr[index];
+    Expects(index >= 0 && index < narrow_cast<decltype(index)>(N));
+    return arr[static_cast<decltype(N)>(index)];
 }
 
 template <class T, size_t N>
-constexpr T& at(std::array<T, N>& arr, size_t index)
+constexpr T& at(std::array<T, N>& arr, std::ptrdiff_t index)
 {
-    Expects(index < N);
-    return arr[index];
+    Expects(index >= 0 && index < narrow_cast<decltype(index)>(N));
+    return arr[static_cast<decltype(N)>(index)];
 }
```

Comment:
😄 

I can certainly change it back, but we use aliases in classes for all of our types for the same reasons (well there are many reasons for it, but this is one of them). Since this is a standalone function you would have to make the aliases global which is not a good idea, so I used decltype() as the next best thing for consistency. 

I have no issue changing it back (it's an easy change), but IMO it seems weird to redefine the type twice. 

I leave it up to you.